### PR TITLE
feat: implement the UI/UX in checkout to place replenishment orders (GH-8473)

### PR DESCRIPTION
**Pull request by bgambocjaviniar**
closes GH-8473

- datepicker will be changed according to what `my-company branch` has for re-usability in a later ticket
- styling for the repeat-days of weekly on the replenishment card or  the whole card in general will be done in another ticket
- Use the process state to create a loading effect for placing an order in another `stretch` ticket

### CODE REVIEW

File: projects/storefrontlib/src/cms-components/checkout/services/checkout-replenishment-form-service.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+  constructor() {}
+
+  getScheduleReplenishmentFormData(): Observable<ScheduleReplenishmentForm> {
+    return this.scheduleReplenishmentFormData$.asObservable();
+  }
+
+  setScheduleReplenishmentFormData(formData: ScheduleReplenishmentForm): void {
+    this.scheduleReplenishmentFormData$.next(formData);
+  }
```
Comment (LTiger14) [inline projects/storefrontlib/src/cms-components/checkout/services/checkout-replenishment-form-service.ts:37]: Maybe you could add a reset method that sets it back to the default state? I remember you doing that somewhere

File: projects/storefrontlib/src/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.component.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+  changeNumberOfDays(nDays: string): void {
+    this.checkoutReplenishmentFormService.setScheduleReplenishmentFormData({
+      ...this.scheduleReplenishmentFormData,
+      numberOfDays: nDays,
+    });
+  }
+
+  changeRecurrencePeriodType(type: string): void {
+    if (type === recurrencePeriod.WEEKLY) {
```
Comment (LTiger14) [inline projects/storefrontlib/src/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.component.ts]: `this.isWeekly = type === recurrencePeriod.WEEKLY`?

File: projects/storefrontlib/src/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.component.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+  hasDaysOfWeekChecked(day: DaysOfWeek) {
+    return this.currentDaysOfWeek.includes(day);
+  }
+
+  private initConfig() {
+    this.isMonthly =
+      this.scheduleReplenishmentFormData.recurrencePeriod ===
+      recurrencePeriod.MONTHLY
+        ? true
+        : false;
```
Comment (LTiger14) [inline projects/storefrontlib/src/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.component.ts]: See comments above. Also add some void types 😉

File: projects/storefrontlib/src/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.component.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+      this.isWeekly = false;
+    }
+
+    if (type === recurrencePeriod.MONTHLY) {
+      this.isMonthly = true;
+    } else {
+      this.isMonthly = false;
+    }
+
+    this.checkoutReplenishmentFormService.setScheduleReplenishmentFormData({
```
Comment (LTiger14) [inline projects/storefrontlib/src/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.component.ts:88]: I like this immutable approach to your service "state".

File: projects/storefrontlib/src/cms-components/checkout/components/place-order/place-order.component.ts [hunk 5]
```diff
@@ -5,48 +5,104 @@ import {
...
+    this.subscription.add(
+      this.checkoutReplenishmentService
+        .getReplenishmentOrderDetails()
+        .pipe(filter((order) => Object.keys(order).length !== 0))
+        .subscribe(() => {
+          this.routingService.go({ cxRoute: 'replenishmentConfirmation' });
+        })
+    );
+
+    this.subscription.add(
```
Comment (LTiger14) [inline projects/storefrontlib/src/cms-components/checkout/components/place-order/place-order.component.ts:84]: Is there a way to get less subscriptions going on? 😅

File: projects/storefrontlib/src/cms-components/checkout/services/checkout-replenishment-form-service.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  /**
+   * Default form data
+   */
+  readonly defaultFormData: ScheduleReplenishmentForm = {
+    daysOfWeek: [],
+    nthDayOfMonth: '1',
+    numberOfDays: '7',
+    numberOfWeeks: '1',
+    recurrencePeriod: recurrencePeriod.DAILY,
+    replenishmentStartDate: new Date().toISOString().split('.')[0] + 'Z',
```
Comment (LTiger14) [inline projects/storefrontlib/src/cms-components/checkout/services/checkout-replenishment-form-service.ts]: Could you use a serializer to format the data correctly to send to the BE?

File: projects/storefrontlib/src/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+    expect(
+      checkoutReplenishmentFormService.setScheduleReplenishmentFormData
+    ).toHaveBeenCalledWith({
+      ...mockReplenishmentOrderFormData,
+      nthDayOfMonth: mockDayOfMonth,
+    });
+  });
+
+  // it('should change replenishment start date of replenishment form data', () => {
```
Comment (LTiger14) [inline projects/storefrontlib/src/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.component.spec.ts]: These tests appear not to run.

**Review by LTiger14 — CHANGES_REQUESTED** [review]
Screen Shot 2020-08-18 at 12 03 36 PM

Was this text always so small?

Quick question, if I select 31st day of the month what happens on a month with 30 days?

**Comment by LTiger14** [comment]
Additional question about the number of weeks drop-down, should it always go to 31?

File: projects/storefrontlib/src/cms-components/checkout/components/place-order/place-order.component.ts (hunk 5 above)
Comment (bgambocjaviniar) [reply projects/storefrontlib/src/cms-components/checkout/components/place-order/place-order.component.ts:84]: maybe

**Comment by bgambocjaviniar** [comment]
> Additional question about the number of weeks drop-down, should it always go to 31?

> Quick question, if I select 31st day of the month what happens on a month with 30 days?

made it dynamic now, so when the date changed, it will change to the number of days accordingly, but this made the complexity of the changeReplDate function funky

File: projects/storefrontlib/src/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.component.ts [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+  }
+
+  changeReplenishmentStartDate(date: string): void {
+    if (Boolean(date)) {
+      this.numberOfDays = this.createNumberStringArray(
+        this.getNumberOfDayInAMonth(date)
+      );
+
+      if (
+        Number(this.scheduleReplenishmentFormData.nthDayOfMonth) >
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.component.ts]: :)

Comment (LTiger14) [reply projects/storefrontlib/src/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.component.ts]: Won't this be handle with the my-company date picker?

Comment (bgambocjaviniar) [reply projects/storefrontlib/src/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.component.ts]: nope don't think so as it doesn't associate with the number of days input and day of the month input

Comment (bgambocjaviniar) [reply projects/storefrontlib/src/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.component.ts]: using defaults as in accelerator as Bill mentioned

File: projects/storefrontstyles/scss/components/checkout/steps/review/_review-submit.scss [hunk 9]
```diff
@@ -153,19 +153,117 @@ cx-place-order {
     --cx-flex-basis: 45%;
   }
 
+  .scaled-input {
+    transform: scale(0.7);
+  }
+
   .cx-place-order-form {
-    @include type('7');
+    font-size: 12px;
```
Comment (LTiger14) [inline projects/storefrontstyles/scss/components/checkout/steps/review/_review-submit.scss:161]: You could add a font type for this like 9

File: projects/core/src/occ/adapters/checkout/converters/occ-replenishment-order-form-serializer.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import { Injectable } from '@angular/core';
+import { ScheduleReplenishmentForm } from '../../../../model/replenishment-order.model';
+import { Converter } from '../../../../util/converter.service';
+import { Occ } from '../../../occ-models/occ.models';
+
+@Injectable({ providedIn: 'root' })
+export class OccReplenishmentOrderFormSerializer
```
Comment (LTiger14) [inline projects/core/src/occ/adapters/checkout/converters/occ-replenishment-order-form-serializer.ts:7]: 🔥

**Review by LTiger14 — APPROVED** [review]

**Pull request merged**